### PR TITLE
[SelectionDAG] Add space-optimized forms of OPC_CheckPredicate

### PR DESCRIPTION
**Pull request by wangpc-pp**
We record the usage of each `Predicate` and sort them by usage.

For the top 8 `Predicate`s, we will emit a `PC_CheckPredicateN` to
save one byte.

Overall this reduces the llc binary size with all in-tree targets by
about 61K.

This PR is stacked on #73310.

### CODE REVIEW

**Comment by wangpc-pp** [comment]
Ping.

File: llvm/utils/TableGen/CodeGenDAGPatterns.h [hunk 1]
```diff
@@ -1163,6 +1169,23 @@ class CodeGenDAGPatterns {
...
+  void increaseComplexPatternUsage(const ComplexPattern *CP) {
+    ComplexPatternUsage[CP]++;
+  }
+
+  const std::map<TreePattern *, unsigned> &getPredicateUsage() const {
+    return PredicateUsage;
+  }
+
+  void increasePredicateUsage(const TreePredicateFn &Predicate) {
+    PredicateUsage[Predicate.getOrigPatFragRecord()]++;
```
Comment (arsenm) [inline llvm/utils/TableGen/CodeGenDAGPatterns.h:1186]: Pre-increment

File: llvm/utils/TableGen/CodeGenDAGPatterns.h [hunk 2]
```diff
@@ -1163,6 +1169,23 @@ class CodeGenDAGPatterns {
     return F->second;
   }
 
+  const std::map<const ComplexPattern *, unsigned> &
+  getComplexPatternUsage() const {
+    return ComplexPatternUsage;
+  }
+
+  void increaseComplexPatternUsage(const ComplexPattern *CP) {
+    ComplexPatternUsage[CP]++;
```
Comment (arsenm) [inline llvm/utils/TableGen/CodeGenDAGPatterns.h:1178]: Pre-increment

File: llvm/utils/TableGen/DAGISelMatcherEmitter.cpp [hunk 3]
```diff
@@ -85,7 +83,45 @@ class MatcherTableEmitter {
...
+    for (auto &Usage : PredicateUsage) {
+      TreePattern *TP = Usage.first;
+      TreePredicateFn Pred(TP);
+      NodePredicatesByCodeToRun[Pred.getCodeToRunOnSDNode()].push_back(TP);
+    }
+
+    std::vector<std::pair<TreePattern *, unsigned>> PredicateList;
+    // Sum the usage.
+    for (auto &Predicate : NodePredicatesByCodeToRun) {
+      TinyPtrVector<TreePattern *> TPs = Predicate.second;
```
Comment (arsenm) [inline llvm/utils/TableGen/DAGISelMatcherEmitter.cpp:107]: Why copy?

**Comment by wangpc-pp** [comment]
Ping.

**Review by arsenm — APPROVED** [review]

**Review by ilovepi — APPROVED** [review]

**Comment by metaflow** [comment]
Test address-space-patfrags.td.test started to fire https://lab.llvm.org/buildbot/#/builders/104/builds/15012

**Comment by metaflow** [comment]
I have reverted this for now

**Comment by wangpc-pp** [comment]
> I have reverted this for now

Thanks! I can't reproduce the failure, I will try to fix it later.

**Comment by RKSimon** [comment]
@wangpc-pp  This is on an expensive checks build (cmake -DLLVM_ENABLE_EXPENSIVE_CHECKS=true) - these builds do extra stable sorting which can sometimes cause check diffs

**Comment by MaskRay** [comment]
One of the recent TableGen changes (likely `DAGISelMatcherEmitter.cpp`) may make output `*GenDAGISel.inc` non-deterministic. Things like DenseMap iterating needs some attention.

File: llvm/utils/TableGen/DAGISelMatcherEmitter.cpp [hunk 4]
```diff
@@ -125,6 +128,39 @@ class MatcherTableEmitter {
          [](const auto &A, const auto &B) { return A.second > B.second; });
     for (const auto &PatternPredicate : PatternPredicateList)
       PatternPredicates.push_back(PatternPredicate.first);
+
+    // Sort Predicates by usage.
+    // Merge predicates with same code.
+    for (const auto &Usage : PredicateUsage) {
+      TreePattern *TP = Usage.first;
+      TreePredicateFn Pred(TP);
+      NodePredicatesByCodeToRun[Pred.getCodeToRunOnSDNode()].push_back(TP);
```
Comment (MaskRay) [inline llvm/utils/TableGen/DAGISelMatcherEmitter.cpp:137]: Iterating over `PredicateUsage` and creating `StringMap NodePredicatesByCodeToRun` can cause non-determinism.

**Pull request merged**